presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -616,22 +638,98 @@ private static void generatePositionEqualsRowWithPageMethod(
+    }
+
+    private void generatePositionNotDistinctFromRowWithPageMethod(
+            ClassDefinition classDefinition,
+            CallSiteBinder callSiteBinder,
+            List<Type> joinChannelTypes,
+            List<FieldDefinition> joinChannelFields)
+    {
+        Parameter leftBlockIndex = arg("leftBlockIndex", int.class);
+        Parameter leftBlockPosition = arg("leftBlockPosition", int.class);
+        Parameter rightPosition = arg("rightPosition", int.class);
+        Parameter page = arg("page", Page.class);
+        Parameter rightChannels = arg("rightChannels", int[].class);
+
+        MethodDefinition positionNotDistinctFromRowMethod = classDefinition.declareMethod(
+                a(PUBLIC),
+                "positionNotDistinctFromRow",
+                type(boolean.class),
+                leftBlockIndex,
+                leftBlockPosition,
+                rightPosition,
+                page,
+                rightChannels);
+
+        Variable thisVariable = positionNotDistinctFromRowMethod.getThis();
+        Scope scope = positionNotDistinctFromRowMethod.getScope();
+        BytecodeBlock body = positionNotDistinctFromRowMethod.getBody();
+        if (groupByUsesEqualTo) {
+            // positionNotDistinctFromRow delegates to positionEqualsRow when groupByUsesEqualTo is set.
+            body.append(thisVariable.invoke(
+                    "positionEqualsRow",
+                    boolean.class,
+                    leftBlockIndex,
+                    leftBlockPosition,
+                    rightPosition,
+                    page,
+                    rightChannels).ret());
+            return;
+        }
+        scope.declareVariable("wasNull", body, constantFalse());
+        for (int index = 0; index < joinChannelTypes.size(); index++) {
+            BytecodeExpression leftBlock = thisVariable
+                    .getField(joinChannelFields.get(index))
+                    .invoke("get", Object.class, leftBlockIndex)
+                    .cast(Block.class);
+            BytecodeExpression rightBlock = page.invoke("getBlock", Block.class, rightChannels.getElement(index));
+            Type type = joinChannelTypes.get(index);
+            // This is a hack for performance reasons.
+            // Type.equalTo takes two pairs of Block+position.
+            // On the other hand, NOT_DISTINCT_FROM is an operator. It takes two Slices.
+            // As a result, two Slices must be constructed for each invocation, which has a nontrivial cost.
+            // For these types, their equal semantics is known to be the same as not-distinct-from except for null values.
+            //
+            // The plan is to allow scalar function to optionally provide an additional implementation using Block+position calling convention.
+            // At that point, we'll be able to fully deprecate Type.equalTo (and friends) and remove this hack.
+            if (type.getJavaType().equals(Slice.class)) {
```

Comment:
I wonder if it shouldn't be the way around?
That is by default use `Type#equals` and only use `NOT_DISTINCT_FROM` for numeric types.
The reasoning is that for most types `equals == NOT_DISTINCT_FROM`. The only case otherwise are some numeric types.
So instead of altering whole world to fix numeric types group by (exotic use case?) make a dedicated fix only for numeric types.
Both approaches are hack, but the latter seems less intrusive. What do you think?

There is still slight but visible perf regression (unless I'm reading something incorrectly):
```
addPagePreCompute | 20 | BIGINT | t | 1511.502 | 1633.207 |  
-- | -- | -- | -- | -- | -- | --
groupByHashPreCompute | 20 | BIGINT | t | 1537.845 | 1666.830
-- | -- | -- | -- | -- | --
```

Reply:
You are definitely reading this correctly. The regression appears more pronounced for BIGINT when hash enabled. Will follow up and get back to you.

Reply:
This also affects structural types. This PR fixes GROUP BY for arrays/maps/rows with nulls in it.

The right thing to do here is to use "not distinct from". I agree that hack is involved in both ways. However, it's pretty different: a hack that only kick in in specified cases != a hack that kicks except in specified cases.

The majority of CPU cost of a query generally goes to table write/read. As a result, I think 3%-6% for join is not something we need to be too concerned about. cc @martint

Reply:
At the moment, we don't find it worthwhile to figure out exactly what caused that 3%-6% regression. However, we want to put in effort to find out whether the regression comes from invoking MethodHandle (instead of invokestatic).

@elonazoulay is going to do that by hacking the benchmark to call Type.equalTo as a MethodHandle (instead of through invokestatic), and see whether performance numbers change.

Reply:
> This also affects structural types. This PR fixes GROUP BY for arrays/maps/rows with nulls in it.

@haozhun How structural types with nulls are grouped and how does this PR fix that?
Correct me if I am wrong, but structural types with null on different positions will have different hash values so they will land on different machines.

The problem with structural types and nulls is much more complex. For structural types to work in `SemiJoin` for instance we would require a new partitioning function (and not a trivial one).